bokehjs/src/lib/models/annotations/arrow.ts

```diff
@@ -60,91 +66,97 @@ export class ArrowView extends AnnotationView {
+      this._sx_start = this.coordinates.x_scale.v_compute(this._x_start)
+      this._sy_start = this.coordinates.y_scale.v_compute(this._y_start)
     } else {
-      sx_start = frame.xview.v_compute(this._x_start)
-      sy_start = frame.yview.v_compute(this._y_start)
+      this._sx_start = frame.xview.v_compute(this._x_start)
+      this._sy_start = frame.yview.v_compute(this._y_start)
     }
 
-    let sx_end, sy_end
-    if (this.model.end_units == 'data') {
-      sx_end = this.coordinates.x_scale.v_compute(this._x_end)
-      sy_end = this.coordinates.y_scale.v_compute(this._y_end)
+    if (this.model.end_units == "data") {
+      this._sx_end = this.coordinates.x_scale.v_compute(this._x_end)
+      this._sy_end = this.coordinates.y_scale.v_compute(this._y_end)
     } else {
-      sx_end = frame.xview.v_compute(this._x_end)
-      sy_end = frame.yview.v_compute(this._y_end)
+      this._sx_end = frame.xview.v_compute(this._x_end)
+      this._sy_end = frame.yview.v_compute(this._y_end)
     }
 
-    return [[sx_start, sy_start], [sx_end, sy_end]]
-  }
-
-  protected _render(): void {
-    const {ctx} = this.layer
-    ctx.save()
-
-    // Order in this function is important. First we draw all the arrow heads.
-    const [start, end] = this._map_data()
+    const {_sx_start, _sy_start, _sx_end, _sy_end} = this
 
-    if (this.end != null)
-      this._arrow_head(ctx, "render", this.end, start, end)
-    if (this.start != null)
-      this._arrow_head(ctx, "render", this.start, end, start)
+    const n = _sx_start.length
+    const angles = this._angles = new NumberArray(n)
 
-    // Next we call .clip on all the arrow heads, inside an initial canvas sized
-    // rect, to create an "inverted" clip region for the arrow heads
-    ctx.beginPath()
-    const {x, y, width, height} = this.plot_view.frame.bbox
-    ctx.rect(x, y, width, height)
-    if (this.end != null)
-      this._arrow_head(ctx, "clip", this.end, start, end)
-    if (this.start != null)
-      this._arrow_head(ctx, "clip", this.start, end, start)
-    ctx.closePath()
-    ctx.clip()
-
-    // Finally we draw the arrow body, with the clipping regions set up. This prevents
-    // "fat" arrows from overlapping the arrow head in a bad way.
-    this._arrow_body(ctx, start, end)
-
-    ctx.restore()
-  }
-
-  protected _arrow_head(ctx: Context2d, action: "render" | "clip", head: ArrowHeadView, start: Coords, end: Coords): void {
-    for (let i = 0, _end = this._x_start.length; i < _end; i++) {
-      // arrow head runs orthogonal to arrow body
-      const angle = Math.PI/2 + atan2([start[0][i], start[1][i]], [end[0][i], end[1][i]])
-
-      ctx.save()
-
-      ctx.translate(end[0][i], end[1][i])
-      ctx.rotate(angle)
-
-      if (action == "render")
-        head.render(ctx, i)
-      else if (action == "clip")
-        head.clip(ctx, i)
-
-      ctx.restore()
+    for (let i = 0; i < n; i++) {
+      // arrow head runs orthogonal to arrow body (???)
```

Comment:
dunno what this comment was for it can probably just be deleted

Reply:
I will clean this up next time I change this file.